File: packages/filters-webgl/src/gradientBevelFilter.ts
import type { WebGLRenderStateInternal, WebGLRenderTarget } from '@flighthq/render-webgl';
import type { GradientBevelFilter } from '@flighthq/types';
import type { WebGLRenderState } from '@flighthq/types';

import { applyBlurFilterToWebGL } from './blurFilter';
import type { WebGLFilterLocations } from './filterPass';
import { clearWebGLFilterTarget, compileWebGLFilterProgram, drawWebGLFilterPass } from './filterPass';
import { createWebGLGradientRampTexture } from './gradientRamp';
import { applyBlitPass, applyTintPass } from './tintShader';

// Samples the blurred alpha at +offset and -offset to compute a bevel value
// in [-1, 1], mapped to [0, 1] for gradient lookup. Outputs the encoded
// bevel value in the red channel; alpha=1 (will be clipped later).
const BEVEL_ENCODE_FRAGMENT_SRC = `#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform vec2 u_offset;
out vec4 fragColor;
void main() {
  float high = texture(u_texture, v_texCoord - u_offset).a;
  float low = texture(u_texture, v_texCoord + u_offset).a;
  float bevelVal = clamp((high - low) * 0.5 + 0.5, 0.0, 1.0);
  fragColor = vec4(bevelVal, 0.0, 0.0, 1.0);
}`;

// Looks up the encoded bevel value (in .r) in the gradient ramp and clips
// the result to the source alpha (unit 1 holds the source).
const BEVEL_APPLY_FRAGMENT_SRC = `#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform sampler2D u_ramp;
uniform sampler2D u_source;
out vec4 fragColor;
void main() {
  float bevelVal = texture(u_texture, v_texCoord).r;
  vec4 color = texture(u_ramp, vec2(bevelVal, 0.5));
  float srcAlpha = texture(u_source, v_texCoord).a;
  fragColor = color * srcAlpha;
}`;

type BevelEncodeLocations = WebGLFilterLocations & {
  locOffset: WebGLUniformLocation;
};

type BevelApplyLocations = WebGLFilterLocations & {
  locRamp: WebGLUniformLocation;
  locSource: WebGLUniformLocation;
};

const encodeShaders = new WeakMap<WebGLRenderState, BevelEncodeLocations>();
const applyShaders = new WeakMap<WebGLRenderState, BevelApplyLocations>();

/**
 * Applies a gradient bevel to `source`, writing the result to `dest`.
 * The gradient maps bevel depth (shadow edge → highlight edge) to colors.
 *
 * `scratch` must contain three render targets of the same dimensions as `dest`.
 * The filter allocates a temporary `WebGLTexture` internally on each call.
 */
export function applyGradientBevelFilterToWebGL(
  state: WebGLRenderState,
  source: WebGLRenderTarget,
  dest: WebGLRenderTarget,
  scratch: WebGLRenderTarget[],
  filter: Readonly<Omit<GradientBevelFilter, 'type'>>,
): void {
  const angle = ((filter.angle ?? 45) * Math.PI) / 180;
  const distance = filter.distance ?? 4;
  const quality = Math.max(1, Math.round(filter.quality ?? 1));
  const strength = filter.strength ?? 1;

  const [s0, s1, s2] = scratch;
  const internal = state as WebGLRenderStateInternal;
  const { gl } = internal;

  // Build blur basis → s1
  applyTintPass(state, source, s0, 0xffffff, 1, Math.min(1, strength));
  applyBlurFilterToWebGL(state, s0, s1, s2, { blurX: filter.blurX ?? 4, blurY: filter.blurY ?? 4, quality });

  // Encode bevel value from blurred alpha offset samples → s0
  const dx = Math.cos(angle) * distance;
  const dy = Math.sin(angle) * distance;
  applyBevelEncodePass(state, s1, s0, dx / s1.width, dy / s1.height);

  // Build gradient ramp, apply to encoded bevel value clipped to source alpha → s1
  const ramp = createWebGLGradientRampTexture(gl, filter.colors, filter.alphas, filter.ratios);
  applyBevelApplyPass(state, s0, ramp, source, s1);
  gl.deleteTexture(ramp);

  clearWebGLFilterTarget(state, dest);
  if (!(filter.bevelType && filter.bevelType !== 'full')) {
    applyBlitPass(state, source, dest);
  }
  applyBlitPass(state, s1, dest);
}

function applyBevelEncodePass(
  state: WebGLRenderState,
  blurred: WebGLRenderTarget,
  dest: WebGLRenderTarget,
  dx: number,
  dy: number,
): void {
  const loc = getEncodeShader(state);
  drawWebGLFilterPass(state, blurred, dest, loc, (gl) => {
    gl.uniform2f(loc.locOffset, dx, dy);
  });
}

function applyBevelApplyPass(
  state: WebGLRenderState,
  encoded: WebGLRenderTarget,
  ramp: WebGLTexture,
  source: WebGLRenderTarget,
  dest: WebGLRenderTarget,
): void {
  const loc = getApplyShader(state);
  drawWebGLFilterPass(state, encoded, dest, loc, (gl) => {
    gl.activeTexture(gl.TEXTURE1);
    gl.bindTexture(gl.TEXTURE_2D, ramp);
    gl.uniform1i(loc.locRamp, 1);
    gl.activeTexture(gl.TEXTURE2);
    gl.bindTexture(gl.TEXTURE_2D, source.texture);
    gl.uniform1i(loc.locSource, 2);
    gl.activeTexture(gl.TEXTURE0);
  });
}

function getEncodeShader(state: WebGLRenderState): BevelEncodeLocations {
  let loc = encodeShaders.get(state);
  if (loc === undefined) {
    const gl = (state as WebGLRenderStateInternal).gl;
    const base = compileWebGLFilterProgram(gl, BEVEL_ENCODE_FRAGMENT_SRC);
    loc = { ...base, locOffset: gl.getUniformLocation(base.program, 'u_offset')! };
    encodeShaders.set(state, loc);
  }
  return loc;
}

function getApplyShader(state: WebGLRenderState): BevelApplyLocations {
  let loc = applyShaders.get(state);
  if (loc === undefined) {
    const gl = (state as WebGLRenderStateInternal).gl;
    const base = compileWebGLFilterProgram(gl, BEVEL_APPLY_FRAGMENT_SRC);
    loc = {
      ...base,
      locRamp: gl.getUniformLocation(base.program, 'u_ramp')!,
      locSource: gl.getUniformLocation(base.program, 'u_source')!,
    };
    applyShaders.set(state, loc);
  }
  return loc;
}

File: crates/flighthq-filters-gl/src/gradient_bevel_filter.rs
//! Gradient-bevel GL filter pass.
//!
//! The gradient maps bevel depth (shadow edge → highlight edge) to colors.
//!
//! `scratch` must contain three render targets of the same dimensions as `dest`.
//! This function allocates a temporary GL texture internally on each call
//! (the gradient ramp); it is deleted before the function returns.

use flighthq_filters::BevelType;
use flighthq_filters::GradientBevelFilter;
use glow::HasContext;

use crate::blur_filter::apply_box_blur_filter_to_gl;
use crate::filter_pass::{
    GlFullscreenProgram, clear_gl_render_target, draw_gl_fullscreen_pass, get_gl_filter_program,
};
use crate::gradient_ramp::create_gl_gradient_ramp_texture;
use crate::tint_shader::{apply_gl_blit_pass, apply_gl_tint_pass};
use crate::{GlRenderState, GlRenderTarget};

// Samples the blurred alpha at +offset and -offset to compute a bevel value in
// [-1, 1], mapped to [0, 1] for gradient lookup. Encoded into the red channel.
const BEVEL_ENCODE_FRAGMENT_SRC: &str = "#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform vec2 u_offset;
out vec4 fragColor;
void main() {
  float high = texture(u_texture, v_texCoord - u_offset).a;
  float low = texture(u_texture, v_texCoord + u_offset).a;
  float bevelVal = clamp((high - low) * 0.5 + 0.5, 0.0, 1.0);
  fragColor = vec4(bevelVal, 0.0, 0.0, 1.0);
}";

// Looks up the encoded bevel value (in .r) in the gradient ramp and clips the
// result to the source alpha (unit 2 holds the source).
const BEVEL_APPLY_FRAGMENT_SRC: &str = "#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform sampler2D u_ramp;
uniform sampler2D u_source;
out vec4 fragColor;
void main() {
  float bevelVal = texture(u_texture, v_texCoord).r;
  vec4 color = texture(u_ramp, vec2(bevelVal, 0.5));
  float srcAlpha = texture(u_source, v_texCoord).a;
  fragColor = color * srcAlpha;
}";

/// Applies a gradient bevel filter to `source`, writing the result to `dest`.
pub fn apply_gradient_bevel_filter_to_gl(
    state: &GlRenderState,
    source: &GlRenderTarget,
    dest: &GlRenderTarget,
    scratch: &[&GlRenderTarget; 3],
    filter: &GradientBevelFilter,
) {
    let angle = filter.angle.unwrap_or(45.0).to_radians();
    let distance = filter.distance.unwrap_or(4.0);
    let quality = filter.quality.unwrap_or(1).max(1);
    let strength = filter.strength.unwrap_or(1.0);

    let s0 = scratch[0];
    let s1 = scratch[1];
    let s2 = scratch[2];

    // Build the blur basis → s1.
    apply_gl_tint_pass(state, source, s0, 0xffffff, 1.0, strength.min(1.0));
    apply_box_blur_filter_to_gl(
        state,
        s0,
        s1,
        s2,
        filter.blur_x.unwrap_or(4.0),
        filter.blur_y.unwrap_or(4.0),
        quality,
    );

    // Encode bevel value from blurred alpha offset samples → s0.
    let dx = angle.cos() * distance;
    let dy = angle.sin() * distance;
    apply_bevel_encode_pass(state, s1, s0, dx / s1.width as f32, dy / s1.height as f32);

    // Build gradient ramp, apply to the encoded bevel clipped to source alpha → s1.
    let ratios: Vec<u8> = filter.ratios.iter().map(|&r| r.round().clamp(0.0, 255.0) as u8).collect();
    let ramp = create_gl_gradient_ramp_texture(state, &filter.colors, &filter.alphas, &ratios);
    apply_bevel_apply_pass(state, s0, ramp, source, s1);
    unsafe {
        state.gl.delete_texture(ramp);
    }

    clear_gl_render_target(state, dest);
    // Full bevel composites the source under the bevel; inner/outer omit the base source.
    if filter.bevel_type.unwrap_or(BevelType::Full) == BevelType::Full {
        apply_gl_blit_pass(state, source, dest);
    }
    apply_gl_blit_pass(state, s1, dest);
}

/// Returns the bevel-apply shader program for `state`, compiling on first use.
pub fn get_gradient_bevel_apply_shader(state: &GlRenderState) -> &GlFullscreenProgram {
    get_gl_filter_program(state, BEVEL_APPLY_FRAGMENT_SRC, |p| &mut p.gradient_bevel_apply)
}

/// Returns the bevel-encode shader program for `state`, compiling on first use.
pub fn get_gradient_bevel_encode_shader(state: &GlRenderState) -> &GlFullscreenProgram {
    get_gl_filter_program(state, BEVEL_ENCODE_FRAGMENT_SRC, |p| &mut p.gradient_bevel_encode)
}

// Encodes the bevel value from offset alpha samples of the blurred basis.
fn apply_bevel_encode_pass(
    state: &GlRenderState,
    blurred: &GlRenderTarget,
    dest: &GlRenderTarget,
    dx: f32,
    dy: f32,
) {
    let program = get_gradient_bevel_encode_shader(state);
    draw_gl_fullscreen_pass(state, program, &[blurred.texture], Some(dest), move |gl, p| unsafe {
        gl.uniform_2_f32(gl.get_uniform_location(p, "u_offset").as_ref(), dx, dy);
    });
}

// Looks up the encoded bevel (unit 0) in the ramp (unit 1), clipped to source
// alpha (unit 2).
fn apply_bevel_apply_pass(
    state: &GlRenderState,
    encoded: &GlRenderTarget,
    ramp: glow::Texture,
    source: &GlRenderTarget,
    dest: &GlRenderTarget,
) {
    let source_texture = source.texture;
    let program = get_gradient_bevel_apply_shader(state);
    draw_gl_fullscreen_pass(state, program, &[encoded.texture], Some(dest), move |gl, p| unsafe {
        gl.active_texture(glow::TEXTURE1);
        gl.bind_texture(glow::TEXTURE_2D, Some(ramp));
        gl.uniform_1_i32(gl.get_uniform_location(p, "u_ramp").as_ref(), 1);
        gl.active_texture(glow::TEXTURE2);
        gl.bind_texture(glow::TEXTURE_2D, Some(source_texture));
        gl.uniform_1_i32(gl.get_uniform_location(p, "u_source").as_ref(), 2);
        gl.active_texture(glow::TEXTURE0);
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    // BEVEL_APPLY_FRAGMENT_SRC

    #[test]
    fn bevel_apply_fragment_src_clips_ramp_to_source_alpha() {
        assert!(BEVEL_APPLY_FRAGMENT_SRC.contains("uniform sampler2D u_source"));
        assert!(BEVEL_APPLY_FRAGMENT_SRC.contains("fragColor = color * srcAlpha"));
    }

    // BEVEL_ENCODE_FRAGMENT_SRC

    #[test]
    fn bevel_encode_fragment_src_maps_bevel_to_red_channel() {
        assert!(BEVEL_ENCODE_FRAGMENT_SRC.contains("(high - low) * 0.5 + 0.5"));
        assert!(BEVEL_ENCODE_FRAGMENT_SRC.contains("fragColor = vec4(bevelVal, 0.0, 0.0, 1.0)"));
    }
}
